Denis_Protsenko.py

```diff
@@ -0,0 +1,218 @@
+    author: str
+    year: int
+
+
+# Клас книги
+class Book(PublicModel):
+    def __init__(self, book_model: BookModel):
+        self.title = book_model.title
+        self.author = book_model.author
+        self.year = book_model.year
+
+    def book_info(self):  # Implementing the abstract method
+        return f"{self.title}, Автор: {self.author}, Рік видання: {self.year}"
+
+    def __eq__(self, other):
+        if not isinstance(other, Book):
+            return False
+        return (self.title, self.author, self.year) == (other.title, other.author, other.year)
+
+    def __hash__(self):
+        return hash((self.title, self.author, self.year))
+
+
+# Клас журналу
+class Magazine(PublicModel):
+    def __init__(self, magazine_model: MagazineModel):
+        self.title = magazine_model.title
+        self.author = magazine_model.author
+        self.year = magazine_model.year
+
+    def book_info(self):  # Implementing the abstract method
+        return f"{self.title}, Автор: {self.author}, Рік видання: {self.year}"
+
+    def __eq__(self, other):
+        if not isinstance(other, Magazine):
+            return False
+        return (self.title, self.author, self.year) == (other.title, other.author, other.year)
+
+    def __hash__(self):
+        return hash((self.title, self.author, self.year))
+
+
+# Декоратор для логування при додаванні нової книги до бібліотеки
+def log_addition(func):
+    def wrapper(self, book_model):
+        if isinstance(book_model, BookModel):
+            print(f"Додаємо: {book_model.title} (Автор: {book_model.author})")
+        elif isinstance(book_model, MagazineModel):
+            print(f"Додаємо: {book_model.title} (Автор: {book_model.author})")
+        return func(self, book_model)
+    return wrapper
+
+
+# Декоратор, який перевіряє наявність книги в бібліотеці перед її видаленням
+def check_book_existence(func):
+    def wrapper(*args, **kwargs):
+        library, _book = args[0], args[1]
+        if _book in library.books:
+            removed_book = func(*args, **kwargs)
+            if removed_book:
+                print(f"\nВидаленo: {removed_book}")
+                return removed_book  # Возвращаем результат удаления
+        else:
+            print(f"\nКнига не знайдена у бібліотеці.")
+            return None  # Возвращаем None, если книга не была найдена
+    return wrapper
+
+
+# Клас бібліотеки
+class Library:
+    def __init__(self):
+        self.books = set()
+        self.books_by_author_dict = defaultdict(list)  # Словник автор: [список книг]
+        print("Бібліотека створена\n")
+
+    # Додавання книги в бібліотеку
+    @log_addition
+    def add_book(self, book_model: Union[BookModel, MagazineModel]):
+        if isinstance(book_model, BookModel):
+            book = Book(book_model)
+        elif isinstance(book_model, MagazineModel):
+            existing_magazine = next((book for book in self.books if book.title == book_model.title and isinstance(book, Magazine)), None)
```

Comment:
Why you did this check if book already exist in different ways for book and magazine?